pkg/deploy/controller/generictrigger/controller.go

```diff
@@ -0,0 +1,199 @@
+	if decoded == nil {
+		// The decoded deployment config will never be nil here but a sanity check
+		// never hurts.
+		return false, nil
+	}
+	ictCount, resolved, canTriggerByImageChange := 0, 0, false
+	var causes []deployapi.DeploymentCause
+
+	// IMAGE CHANGE TRIGGERS
+	for _, t := range config.Spec.Triggers {
+		if t.Type != deployapi.DeploymentTriggerOnImageChange {
+			continue
+		}
+		ictCount++
+
+		// If this is the initial deployment then we need to wait for the image change controller
+		// to resolve the image inside the pod template.
+		lastTriggered := t.ImageChangeParams.LastTriggeredImage
+		if len(lastTriggered) == 0 {
+			continue
+		}
+		resolved++
+
+		// Non-automatic triggers should not be able to trigger deployments.
+		if !t.ImageChangeParams.Automatic {
+			continue
+		}
+
+		// We need stronger checks in order to validate that this template
+		// change is an image change. Look at the deserialized config's
+		// triggers and compare with the present trigger.
+		if !triggeredByDifferentImage(*t.ImageChangeParams, *decoded) {
+			continue
+		}
+
+		canTriggerByImageChange = true
+		causes = append(causes, deployapi.DeploymentCause{
+			Type: deployapi.DeploymentTriggerOnImageChange,
+			ImageTrigger: &deployapi.DeploymentCauseImageTrigger{
+				From: kapi.ObjectReference{
+					Name:      t.ImageChangeParams.From.Name,
+					Namespace: t.ImageChangeParams.From.Namespace,
+					Kind:      "ImageStreamTag",
+				},
+			},
+		})
+	}
+
+	// We need to wait for all images to resolve before triggering a new deployment.
+	if ictCount != resolved {
+		return false, nil
+	}
+
+	// CONFIG CHANGE TRIGGERS
+	canTriggerByConfigChange := false
+	// Our deployment config has a config change trigger and no image change has triggered.
+	// If an image change had happened, it would be enough to start a new deployment without
+	// caring about the config change trigger.
+	if deployutil.HasChangeTrigger(config) && !canTriggerByImageChange {
+		// This is the initial deployment or the config has a template change. We need to
+		// kick a new deployment.
+		if config.Status.LatestVersion == 0 || !kapi.Semantic.DeepEqual(config.Spec.Template, decoded.Spec.Template) {
+			canTriggerByConfigChange = true
+			causes = []deployapi.DeploymentCause{{Type: deployapi.DeploymentTriggerOnConfigChange}}
+		}
+	}
+
+	return canTriggerByConfigChange || canTriggerByImageChange, causes
+}
+
+// triggeredByDifferentImage compares the provided image change parameters with those found in the
+// previous deployment config (the one we decoded from the annotations of its latest deployment)
+// and returns whether the two deployment configs have been triggered by a different image change.
+func triggeredByDifferentImage(ictParams deployapi.DeploymentTriggerImageChangeParams, previous deployapi.DeploymentConfig) bool {
+	for _, t := range previous.Spec.Triggers {
+		if t.Type != deployapi.DeploymentTriggerOnImageChange {
```

Comment:
Oh, you're calling this in a loop outside of this method.  Probably ok, unless there are duplicate trigger params in the list.    Which we probably should normalize otherwise